pkg/sql/crdb_internal.go

```diff
@@ -1733,3 +1741,223 @@ CREATE TABLE crdb_internal.partitions (
+			}
+
+			var env []json.JSON
+			for _, v := range n.Env {
+				env = append(attrs, json.FromString(v))
+			}
+
+			activity := json.NewBuilder()
+			for nodeID, values := range n.Activity {
+				b := json.NewBuilder()
+				b.Add("incoming", json.FromInt64(values.Incoming))
+				b.Add("outgoing", json.FromInt64(values.Outgoing))
+				b.Add("latency", json.FromInt64(values.Latency))
+				activity.Add(nodeID.String(), b.Build())
+			}
+
+			if err := addRow(
+				tree.NewDInt(tree.DInt(n.Desc.NodeID)),
+				tree.NewDString(n.Desc.Address.NetworkField),
+				tree.NewDString(n.Desc.Address.AddressField),
+				tree.NewDJSON(json.FromArrayOfJSON(attrs)),
+				tree.NewDJSON(locality.Build()),
+				tree.NewDString(n.Desc.ServerVersion.String()),
+				tree.NewDString(n.BuildInfo.GoVersion),
+				tree.NewDString(n.BuildInfo.Tag),
+				tree.NewDString(n.BuildInfo.Time),
+				tree.NewDString(n.BuildInfo.Revision),
+				tree.NewDString(n.BuildInfo.CgoCompiler),
+				tree.NewDString(n.BuildInfo.Platform),
+				tree.NewDString(n.BuildInfo.Distribution),
+				tree.NewDString(n.BuildInfo.Type),
+				tree.NewDString(dependencies),
+				tree.NewDInt(tree.DInt(n.StartedAt)),
+				tree.NewDInt(tree.DInt(n.UpdatedAt)),
+				tree.NewDJSON(metrics.Build()),
+				tree.NewDJSON(json.FromArrayOfJSON(args)),
+				tree.NewDJSON(json.FromArrayOfJSON(env)),
+				tree.NewDJSON(activity.Build()),
+			); err != nil {
+				return err
+			}
+		}
+		return nil
+	},
+}
+
+// crdbInternalKVStoreStatusTable exposes an information about the cluster stores.
+var crdbInternalKVStoreStatusTable = virtualSchemaTable{
+	schema: `
+CREATE TABLE crdb_internal.kv_store_status (
+  node_id            INT NOT NULL,
+  store_id           INT NOT NULL,
+  attrs              JSON NOT NULL,
+  capacity           INT NOT NULL,
+  available          INT NOT NULL,
+  used               INT NOT NULL,
+  logical_bytes      INT NOT NULL,
+  range_count        INT NOT NULL,
+  lease_count        INT NOT NULL,
+  writes_per_second  FLOAT NOT NULL,
+  bytes_per_replica  JSON NOT NULL,
+  writes_per_replica JSON NOT NULL,
+  metrics            JSON NOT NULL
+)
+	`,
+	populate: func(ctx context.Context, p *planner, prefix string, addRow func(...tree.Datum) error) error {
+		if err := p.RequireSuperUser("read crdb_internal.kv_store_status"); err != nil {
+			return err
+		}
+
+		response, err := p.ExecCfg().StatusServer.Nodes(ctx, &serverpb.NodesRequest{})
+		if err != nil {
+			return err
+		}
+
+		for _, n := range response.Nodes {
+			for _, s := range n.StoreStatuses {
+				var attrs []json.JSON
+				for _, a := range s.Desc.Attrs.Attrs {
+					attrs = append(attrs, json.FromString(a))
+				}
+
+				metrics := json.NewBuilder()
+				for k, v := range s.Metrics {
+					metric, err := json.FromFloat64(v)
+					if err != nil {
+						return err
+					}
+					metrics.Add(k, metric)
+				}
+
+				percentilesToJSON := func(ps roachpb.Percentiles) (json.JSON, error) {
```

Comment:
@justinj do you think this code is good as is?

Reply:
Assuming it's reasonable that these columns are JSON in the first place (based on a quick glance it looks like it is) the code constructing them looks good to me!